### PR TITLE
Fix issue where full-SHA alias is not supported for links with user defined networks

### LINKED ISSUE #35573: Links in user defined networks don't support full-SHA aliasing
**Description**

Docker create container, in a user defined network, when providing `--link` doesn't support the forwarding of a full container ID SHA to an alias. Yet it does support the short form.

**Steps to reproduce the issue:**
1. `docker network create foo`
2. `docker run -dit --rm --net "foo" -p "8000" python python3 -m http.server` # e.g. > f10b8d3ec65690692306e7825284603ad648b6610ec8ecf5df97e5921616a619
3. `docker run -it --rm --net "foo" --link "f10b8d3ec65690692306e7825284603ad648b6610ec8ecf5df97e5921616a619:server" ubuntu apt-get update && apt-get install -y curl && curl server:8000` # hostname not found
4. `docker run -it --rm --net "foo" --link "f10b8d3ec656:server" ubuntu apt-get update && apt-get install -y curl && curl server:8000` # spits out the root directory contents of f10b8d3ec656

**Describe the results you received:**

As the comments describe above. The linked python container is NOT reachable in step 3, where it is reachable in step 4.

You will need to substitute in the returned container ID from step 2 in to both step 3 (in full) and step 4 (in short).

**Describe the results you expected:**

My assumption would be that the container identifier provided to link could be in the form of `<name|short-SHA|long-SHA>` but I can't actually find any specific documentation which defines this.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.11.0-ce-rc4
 API version:  1.34
 Go version:   go1.8.4
 Git commit:   587f1f0
 Built:        Thu Nov 16 01:17:06 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.11.0-ce-rc4
 API version:  1.34 (minimum version 1.12)
 Go version:   go1.8.5
 Git commit:   587f1f0
 Built:        Thu Nov 16 01:43:23 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 3
 Running: 2
 Paused: 0
 Stopped: 1
Images: 154
Server Version: 17.11.0-ce-rc4
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 992280e8e265f491f7a624ab82f3e238be086e49
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.60-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: XL3E:73IM:YSPH:M4II:UYES:MMDY:AWFB:QQFF:ZKUI:KLVC:6RMW:RBWU
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 41
 Goroutines: 69
 System Time: 2017-11-22T13:13:20.765955985Z
 EventsListeners: 2
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details:**

Hyperkit / Docker for Mac

### PR DESCRIPTION
**Pull request by yongtang**
**- What I did**

This fix tries to address the issue raised in #35573 where full-SHA alias is not supported for links with user defined networks.

**- How I did it**

The reason was that CreateOptionAlias used name directly but CreateOptionMyAlias used short ID of the container to alias.

This fix use short ID for CreateOptionAlias so that the issue could be addressed.

**- How to verify it**

A test case has been added to cover the changes.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

unnamed

This fix fixes #35573.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like there's failures in `docker-py` tests;

```
02:37:42 =================================== FAILURES ===================================
02:37:42 _____________________ TestNetworks.test_connect_with_links _____________________
02:37:42 /docker-py/tests/integration/api_network_test.py:338: in test_connect_with_links
02:37:42     links=[('docker-py-test-upstream', 'bar')])
02:37:42 /docker-py/docker/utils/decorators.py:19: in wrapped
02:37:42     return f(self, resource_id, *args, **kwargs)
02:37:42 /docker-py/docker/utils/decorators.py:34: in wrapper
02:37:42     return f(self, *args, **kwargs)
02:37:42 /docker-py/docker/api/network.py:253: in connect_container_to_network
02:37:42     self._raise_for_status(res)
02:37:42 /docker-py/docker/api/client.py:224: in _raise_for_status
02:37:42     raise create_api_error_from_http_exception(e)
02:37:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:37:42     raise cls(e, response=response, explanation=explanation)
02:37:42 E   NotFound: 404 Client Error: Not Found ("No such container: docker-py-test-upstream")
02:37:42 _____________________ TestNetworks.test_create_with_links ______________________
02:37:42 /docker-py/tests/integration/api_network_test.py:303: in test_create_with_links
02:37:42     links=[('docker-py-test-upstream', 'bar')],
02:37:42 /docker-py/tests/integration/base.py:117: in create_and_start
02:37:42     self.client.start(container)
02:37:42 /docker-py/docker/utils/decorators.py:19: in wrapped
02:37:42     return f(self, resource_id, *args, **kwargs)
02:37:42 /docker-py/docker/api/container.py:1085: in start
02:37:42     self._raise_for_status(res)
02:37:42 /docker-py/docker/api/client.py:224: in _raise_for_status
02:37:42     raise create_api_error_from_http_exception(e)
02:37:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:37:42     raise cls(e, response=response, explanation=explanation)
02:37:42 E   NotFound: 404 Client Error: Not Found ("No such container: docker-py-test-upstream")
```

File: integration/network/link_test.go
Comment (thaJeztah) [inline integration/network/link_test.go]: can you remove this one?

Comment (yongtang) [reply integration/network/link_test.go]: Thanks @thaJeztah. The PR has been updated.

**Comment by yongtang** [comment]
@thaJeztah The PR has been updated. Please take a look.

File: daemon/container_operations.go
Comment (MHBauer) [inline daemon/container_operations.go]: The comment makes it look like these three lines are preserving the old behavior, but this is the implementation of the fix that turns a `name`, which coming out of `ParseLink` is really a full long ID.

I don't even remotely know how libnetwork works to suggest how the comment should be adjusted. Maybe remove the content before the comma, and leave only something like "use truncated container id if linked container was a name"

**Review by MHBauer — COMMENTED** [review]
I am having a bit of trouble with the comments, but I think the code looks okay.

I would like to see additional test cases beyond full ID to the other cases of name and shrort ID, unless those are tested elsewhere.

Remind me to come back around and look at this again.

**Comment by yongtang** [comment]
@MHBauer Thanks for the review. The PR has been updated with additional tests as well as the updated comments. Please take a look.

File: integration/network/link_test.go
Comment (MHBauer) [inline integration/network/link_test.go]: nit,add comment
// these are the test cases of full id,partial id, and container name.

**Comment by MHBauer** [comment]
looks good to me, do we need @dnephin or @vdemeester to take a look? I've never interacted with code owners before.

File: integration/network/link_test.go
Comment (yongtang) [reply integration/network/link_test.go]: @MHBauer Thanks. The comment has been updated.

**Comment by yongtang** [comment]
Ping @dnephin  @vdemeester, please take a look.

File: daemon/container_operations.go
Comment (dnephin) [inline daemon/container_operations.go]: This could still be a separate function in this package.

File: integration/network/link_test.go
Comment (dnephin) [inline integration/network/link_test.go]: It should be possible to test this behaviour without an integration test

**Comment by dnephin** [comment]
code owners are just being used as a way to get notified when a PR changes some file that matches a pattern. It doesn't necessary mean the PR needs approval from one.

**Comment by yongtang** [comment]
@dnephin It might be difficult to come up with a unit test as:
https://github.com/yongtang/docker/blob/9fc57f1c48614668668fa3290cdcf10ee9fb1830/daemon/container_operations.go#L787-L791
```
			// To preseve old behavior, use truncated container id
			// if linked container was a name.
			if child, err := daemon.GetContainer(name); err == nil {
				name = stringid.TruncateID(child.ID)
			}
```
needs to interact with daemon (`daemon.GetContainer(name)`) to resolve the linked container. Not sure what is the best way to handle this case.

**Comment by dnephin** [comment]
If you extract the code back into a function, the function could accept an interface with `GetContainer()`. In the test you can pass a fake that implements the interface to return the value you want for that case.

**Comment by yongtang** [comment]
Thanks @dnephin After taking another look I think the function parameter of `GetContainer` will help.

But there are still other issues. At the moment the
```
BuildJoinOptions(n named) ([]libnetwork.EndpointOption, error) {
```
returns `EndpointOption` which is an option setter function type with unexpected parameter types:
```
type EndpointOption func(ep *endpoint)
```
It is really quite difficult to test against `EndpointOption` it looks.

**Comment by dnephin** [comment]
Ah yes, that does make it very difficult to test. I'm out of ideas

**Comment by yongtang** [comment]
@dnephin Maybe we add the integration test for now, to have the fix for #35573? It may not be ideal though that might be an option at the moment.

**Comment by vdemeester** [comment]
ping @yongtang what is the status here ? also need a rebase :sweat_smile:

**Comment by yongtang** [comment]
Thanks @vdemeester. The PR has been rebased. The integration test is included. Please take a look.

**Review by vdemeester — APPROVED** [review]
SGTM :camel: 
cc @thaJeztah @fcrisciani

**Comment by thaJeztah** [comment]
Looks like a function needs to be removed (or the new code put in that function);

```
07:54:44 daemon/network.go:900:1:warning: buildJoinOptions is unused (deadcode)
07:54:44 Build step 'Execute shell' marked build as failure
```

**Comment by yongtang** [comment]
@vdemeester @thaJeztah The tests passed except powerpc which I think is unrelated (#37277).
```
02:40:23 FAIL: docker_cli_daemon_test.go:1800: DockerDaemonSuite.TestDaemonNoSpaceLeftOnDeviceError
02:40:23 
02:40:23 [dcfb192c0bd25] waiting for daemon to start
02:40:23 [dcfb192c0bd25] exiting daemon
02:40:23 docker_cli_daemon_test.go:1817:
02:40:23     s.d.Start(c, "--data-root", filepath.Join(testDir, "test-mount"))
02:40:23 /go/src/github.com/docker/docker/internal/test/daemon/daemon.go:203:
02:40:23     t.Fatalf("Error starting daemon with arguments: %v", args)
02:40:23 ... Error: Error starting daemon with arguments: [--data-root /tmp/no-space-left-on-device-test032730539/test-mount]
```

**Comment by yongtang** [comment]
@thaJeztah @vdemeester All tests passed now. Please take a look.

**Review by vdemeester — APPROVED** [review]
Still LGTM :stuck_out_tongue:

**Comment by olljanat** [comment]
@thaJeztah PTAL

**Comment by piyush-insider** [comment]
Hi, @yongtang @vdemeester is this being actively worked on? Is it safe to close this PR now?

**Comment by yongtang** [comment]
@thaJeztah Is an integration test OK for this PR? I can rebase and update the PR if this is the case.

**Comment by akerouanton** [comment]
As we're in the process of fully removing legacy links (see this RFC: https://github.com/moby/moby/issues/47464), let me close this PR.

**Pull request closed without merging**
